pkg/kn/commands/plugin/plugin_list.go

```diff
@@ -0,0 +1,190 @@
+	"github.com/spf13/cobra"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+)
+
+// ValidPluginFilenamePrefixes controls the prefix for all kn plugins
+var ValidPluginFilenamePrefixes = []string{"kn"}
+
+// NewPluginListCommand creates a new `kn plugin list` command
+func NewPluginListCommand(p *commands.KnParams) *cobra.Command {
+	pluginFlags := PluginFlags{
+		IOStreams: genericclioptions.IOStreams{
+			In:     os.Stdin,
+			Out:    os.Stdout,
+			ErrOut: os.Stderr,
+		},
+	}
+
+	pluginListCommand := &cobra.Command{
+		Use:   "list",
+		Short: "List all visible plugin executables",
+		Long: `List all visible plugin executables.
+
+		Available plugin files are those that are:
+		- executable
+		- begin with "kn-
+		- anywhere on the path specfied in Kn's config pluginDir variable, which:
+		  * defaults to $PATH if not specified
+		  * can be overridden with the --plugin-dir flag`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			err := pluginFlags.complete(cmd)
+			if err != nil {
+				return err
+			}
+
+			err = pluginFlags.run()
+			if err != nil {
+				return err
+			}
+
+			return nil
+		},
+	}
+
+	pluginFlags.AddPluginFlags(pluginListCommand)
+
+	return pluginListCommand
+}
+
+// Private
+
+func (o *PluginFlags) complete(cmd *cobra.Command) error {
+	o.Verifier = &CommandOverrideVerifier{
+		Root:        cmd.Root(),
+		SeenPlugins: make(map[string]string, 0),
+	}
+
+	pluginPath := commands.PluginDir
+	if pluginPath == "$PATH" {
+		pluginPath = os.Getenv("PATH")
+	}
+
+	o.PluginPaths = filepath.SplitList(pluginPath)
+
+	return nil
+}
+
+func (o *PluginFlags) run() error {
+	pluginsFound := false
+	isFirstFile := true
+	pluginErrors := []error{}
+	pluginWarnings := 0
+
+	for _, dir := range uniquePathsList(o.PluginPaths) {
+		if dir == "" {
+			continue
+		}
+		files, err := ioutil.ReadDir(dir)
+		if err != nil {
+			if _, ok := err.(*os.PathError); ok {
+				fmt.Fprintf(o.ErrOut, "Unable read directory %q from your pluginDir: %v. Skipping...", dir, err)
+				continue
+			}
+
+			pluginErrors = append(pluginErrors, fmt.Errorf("error: unable to read directory %q in your pluginDir: %v", dir, err))
+			continue
+		}
+
+		for _, f := range files {
+			if f.IsDir() {
+				continue
+			}
+			if !hasValidPrefix(f.Name(), ValidPluginFilenamePrefixes) {
+				continue
+			}
+
+			if isFirstFile {
+				fmt.Fprintf(o.ErrOut, "The following compatible plugins are available:\n\n")
+				pluginsFound = true
+				isFirstFile = false
+			}
+
+			pluginPath := f.Name()
+			if !o.NameOnly {
+				pluginPath = filepath.Join(dir, pluginPath)
+			}
+
+			fmt.Fprintf(o.Out, "%s\n", pluginPath)
+			if errs := o.Verifier.Verify(filepath.Join(dir, f.Name())); len(errs) != 0 {
+				for _, err := range errs {
+					fmt.Fprintf(o.ErrOut, "  - %s\n", err)
+					pluginWarnings++
+				}
+			}
+		}
+	}
+
+	if !pluginsFound {
+		pluginErrors = append(pluginErrors, fmt.Errorf("warning: unable to find any kn plugins in your pluginDir"))
```

Comment:
This is not an error

Reply:
yes, there is no way to return a warning error so the message is "warning"